include/StreamBase.h

```diff
@@ -0,0 +1,230 @@
+        /// @brief Gets whether or not the source of this Stream is encrypted.
+        /// @return Returns true if the original data being streamed is encrypted, false otherwise.
+        [[nodiscard]] virtual Boole IsEncrypted() const = 0;
+        /// @brief Gets whether or not this Stream is decompressing/decrypting the source data.
+        /// @remarks A Raw Stream is fairly common for file system and memory streams.  Even with Zip or other
+        /// similar archive streams it can be useful if custom operations are desired, but less common.
+        /// @return Returns true if the source data is being streamed without being altered, false otherwise.
+        [[nodiscard]] virtual Boole IsRaw() const = 0;
+
+        /// @brief Gets whether or not the current position is at the end of the Stream.
+        /// @return Returns true if the current position has reached the end of the Stream, false otherwise.
+        [[nodiscard]] virtual Boole EoF() const = 0;
+        /// @brief Gets whether or not a critical error was detected in a previous operation in the Stream.
+        /// @return Returns true if a critical error has occurred and the Stream integrity may be invalid, false otherwise.
+        [[nodiscard]] virtual Boole Bad() const = 0;
+        /// @brief Gets whether or not an otherwise silent and recoverable error was detected in a previous operation in the Stream.
+        /// @return Returns true if a non-critical operation failed (such as seek), false otherwise.
+        [[nodiscard]] virtual Boole Fail() const = 0;
+        /// @brief Gets whether or not this Stream is intact and ready for operations.
+        /// @return Returns true if no failures have been detected, false otherwise.
+        [[nodiscard]] virtual Boole IsValid() const = 0;
+
+        /// @brief Clears any stored error state on the Stream.
+        /// @remarks This is useful for non-critical errors such as ones that cause "EoF()" or "Fail()" to return true but not
+        /// "Bad()".  Using this to clear critical errors is not advised.
+        virtual void ClearErrors() = 0;
+    };//iStreamBase
+
+    #ifdef _MSC_VER
+    #pragma vtordisp(push, 2)
+    #endif
+
+    ///////////////////////////////////////////////////////////////////////////////
+    /// @brief Interface class for input (read) Streams.
+    ///////////////////////////////////////
+    class iInStream : virtual public iStreamBase
+    {
+    public:
+        /// @brief Class constructor.
+        iInStream() = default;
+        /// @brief Class destructor.
+        virtual ~iInStream() = default;
+
+        /// @brief Reads from the Stream and copies that data to a buffer.
+        /// @param Buffer The buffer to be populated with the read data.
+        /// @param Size The number of bytes to read from the Stream.
+        /// @return Returns the number of bytes successfully read.
+        virtual size_t Read(void* Buffer, const StreamSize Size) = 0;
```

Comment:
These kinds of APIs always rub me the wrong the wrong way. They are error prone and not easily composable. I am glad there is a GetAsString() member that always this to be used without the intentional passing in of a mutable item. I know this is one of the higher performance options, but there are ways to keep that performance that are at least less error prone, like the BinaryBuffer.

We should discuss alternatives. I know this is a point of contention we often have.

Reply:
We discussed this. For others reading this, my takeaway was that there would be simplification here, removing context specific functions. These helper functions would move to additional classes like TextStreamReader and BinaryStreamReader classes that would have context specific helper functions.

Stuff like ReadLine() and GetAsText() could go onto the TextStreamReader.